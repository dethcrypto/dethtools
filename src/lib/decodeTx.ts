import { TransactionFactory, TypedTransaction } from '@ethereumjs/tx';
import { toBuffer } from 'ethereumjs-util';

export function decodeTx(rawTx: string): DecodedTx | undefined {
  try {
    const txDataBuffer = toBuffer(rawTx);
    const tx = TransactionFactory.fromSerializedData(txDataBuffer);
    const senderAddr = tx.getSenderAddress().toString();
    return { tx, senderAddr };
  } catch (error) {
    throw new Error(
      `Failed to decode transaction: ${(error as Error).message}`,
    );
  }
}

export interface DecodedTx {
  tx: TypedTransaction;
  senderAddr: string;
}
