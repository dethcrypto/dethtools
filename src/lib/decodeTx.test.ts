import {
  AccessListEIP2930Transaction,
  FeeMarketEIP1559Transaction,
  Transaction,
} from '@ethereumjs/tx';
import { expect } from 'earljs';

import { decodeTx } from './decodeTx';

describe(decodeTx.name, () => {
  it('handles legacy tx input', () => {
    const actual = decodeTx(
      '0xf86c0a85046c7cfe0083016dea94d1310c1e038bc12865d3d3997275b3e4737c6302880b503be34d9fe80080269fc7eaaa9c21f59adf8ad43ed66cf5ef9ee1c317bd4d32cd65401e7aaca47cfaa0387d79c65b90be6260d09dcfb780f29dd8133b9b1ceb20b83b7e442b4bfc30cb',
    );

    expect(actual.tx).toBeA(Transaction);
    expect(actual.tx).toMatchSnapshot();
    expect(actual.senderAddr).toEqual(
      '0x67835910d32600471f388a137bbff3eb07993c04',
    );
  });

  it('handles EIP-1559 tx input', () => {
    const actual = decodeTx(
      '0x02f87301398459682f0085071c42979e82520894c9e7a3937cf24ed932b69c4545a20a5966a73fb58803311fc80a57000080c001a005a5ee2b0994acdd28f43b8fc2594cd702ada33f1f319308352e401e9b493ef9a011dd2830018dbf05a7f37860d6cff9f530b073836985005083469a86d805a7a2',
    );

    expect(actual.tx).toBeA(FeeMarketEIP1559Transaction);
    expect(actual.tx).toMatchSnapshot();
    expect(actual.senderAddr).toEqual(
      '0xa784398ad2a43b8ba13b0eb64664778f088bf6d2',
    );
  });

  it('handles EIP2930 transaction', () => {
    const validAddress = Buffer.from('01'.repeat(20), 'hex');
    const validSlot = Buffer.from('01'.repeat(32), 'hex');
    const chainId = 2;
    const data = Buffer.from('010200', 'hex');

    const inputTx = AccessListEIP2930Transaction.fromTxData({
      data,
      to: validAddress,
      accessList: [[validAddress, [validSlot]]],
      chainId,
    }).sign(Buffer.from('01'.repeat(32), 'hex'));

    const actual = decodeTx('0x' + inputTx.serialize().toString('hex'));

    expect(actual.tx).toBeA(AccessListEIP2930Transaction);

    const tx = actual.tx as AccessListEIP2930Transaction;

    expect(tx.accessList).toEqual([[validAddress, [validSlot]]]);
    expect(tx.to!.toBuffer()).toEqual(validAddress);
    expect(tx.chainId.toNumber()).toEqual(chainId);
    expect(tx.data).toEqual(data);
    expect(tx).toMatchSnapshot();
  });
});
