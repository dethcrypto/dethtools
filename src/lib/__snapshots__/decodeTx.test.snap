// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decodeTx handles EIP-1559 tx input 1`] = `
Object {
  "accessList": Array [],
  "chainId": "0x1",
  "data": "0x",
  "gasLimit": "0x5208",
  "maxFeePerGas": "0x71c42979e",
  "maxPriorityFeePerGas": "0x59682f00",
  "nonce": "0x39",
  "r": "0x5a5ee2b0994acdd28f43b8fc2594cd702ada33f1f319308352e401e9b493ef9",
  "s": "0x11dd2830018dbf05a7f37860d6cff9f530b073836985005083469a86d805a7a2",
  "to": "0xc9e7a3937cf24ed932b69c4545a20a5966a73fb5",
  "v": "0x1",
  "value": "0x3311fc80a570000",
}
`;

exports[`decodeTx handles EIP2930 transaction 1`] = `
Object {
  "accessList": Array [
    Object {
      "address": "0x0101010101010101010101010101010101010101",
      "storageKeys": Array [
        "0x0101010101010101010101010101010101010101010101010101010101010101",
      ],
    },
  ],
  "chainId": "0x2",
  "data": "0x010200",
  "gasLimit": "0x0",
  "gasPrice": "0x0",
  "nonce": "0x0",
  "r": "0xbb8dbe4eff3c08f6bf5113cbe1b362417d8871c272a766f492d8c5ef7dc91fd3",
  "s": "0xf7c002813490c3ca8afbb779a9ffd2aad4e0e980b55212f1584cf39fe7576c4",
  "to": "0x0101010101010101010101010101010101010101",
  "v": "0x0",
  "value": "0x0",
}
`;

exports[`decodeTx handles legacy tx input 1`] = `
Object {
  "data": "0x",
  "gasLimit": "0x16dea",
  "gasPrice": "0x46c7cfe00",
  "nonce": "0xa",
  "r": "0xc7eaaa9c21f59adf8ad43ed66cf5ef9ee1c317bd4d32cd65401e7aaca47cfa",
  "s": "0x387d79c65b90be6260d09dcfb780f29dd8133b9b1ceb20b83b7e442b4bfc30cb",
  "to": "0xd1310c1e038bc12865d3d3997275b3e4737c6302",
  "v": "0x26",
  "value": "0xb503be34d9fe800",
}
`;
